src/ring_theory/roots_of_unity.lean

```diff
@@ -0,0 +1,586 @@
+
+end comm_group
+
+section comm_group_with_zero
+
+variables {ζ : G₀} (h : is_primitive_root ζ k)
+
+lemma fpow_eq_one : ζ ^ (k : ℤ) = 1 := h.pow_eq_one
+
+lemma fpow_eq_one_iff_dvd (h : is_primitive_root ζ k) (l : ℤ) :
+  ζ ^ l = 1 ↔ (k : ℤ) ∣ l :=
+begin
+  by_cases h0 : 0 ≤ l,
+  { lift l to ℕ using h0, rw [fpow_coe_nat], norm_cast, exact h.pow_eq_one_iff_dvd l },
+  { have : 0 ≤ -l, { simp only [not_le, neg_nonneg] at h0 ⊢, exact le_of_lt h0 },
+    lift -l to ℕ using this with l' hl',
+    rw [← dvd_neg, ← hl'],
+    norm_cast,
+    rw [← h.pow_eq_one_iff_dvd, ← inv_inj', ← fpow_neg, ← hl', fpow_coe_nat, inv_one] }
+end
+
+lemma inv' (h : is_primitive_root ζ k) : is_primitive_root ζ⁻¹ k :=
+{ pow_eq_one := by simp only [h.pow_eq_one, inv_one, eq_self_iff_true, inv_pow'],
+  dvd_of_pow_eq_one :=
+  begin
+    intros l hl,
+    apply h.dvd_of_pow_eq_one l,
+    rw [← inv_inj', ← inv_pow', hl, inv_one]
+  end }
+
+@[simp] lemma inv_iff' : is_primitive_root ζ⁻¹ k ↔ is_primitive_root ζ k :=
+by { refine ⟨_, λ h, inv' h⟩, intro h, rw [← inv_inv' ζ], exact inv' h }
+
+lemma fpow_of_gcd_eq_one (h : is_primitive_root ζ k) (i : ℤ) (hi : i.gcd k = 1) :
+  is_primitive_root (ζ ^ i) k :=
+begin
+  by_cases h0 : 0 ≤ i,
+  { lift i to ℕ using h0, exact h.pow_of_coprime i hi },
+  have : 0 ≤ -i, { simp only [not_le, neg_nonneg] at h0 ⊢, exact le_of_lt h0 },
+  lift -i to ℕ using this with i' hi',
+  rw [← inv_iff', ← fpow_neg, ← hi'],
+  apply h.pow_of_coprime,
+  rw [int.gcd, ← int.nat_abs_neg, ← hi'] at hi,
+  exact hi
+end
+
+end comm_group_with_zero
+
+section integral_domain
+
+variables {ζ : R}
+
+lemma neg_one (p : ℕ) [char_p R p] (hp : p ≠ 2) : is_primitive_root (-1 : R) 2 :=
+mk_of_lt (-1 : R) dec_trivial (by simp only [one_pow, neg_square]) $
+begin
+  intros l hl0 hl2,
+  obtain rfl : l = 1,
+  { unfreezingI { clear_dependent R p }, dec_trivial! },
+  simp only [pow_one, ne.def],
+  intro h,
+  suffices h2 : p ∣ 2,
+  { have := char_p.char_ne_one R p,
+    unfreezingI { clear_dependent R },
+    have aux := nat.le_of_dvd dec_trivial h2,
+    revert this hp h2, revert p, dec_trivial },
+  simp only [← char_p.cast_eq_zero_iff R p, nat.cast_bit0, nat.cast_one],
+  rw [bit0, ← h, neg_add_self] { occs := occurrences.pos [1] }
+end
+
+lemma eq_neg_one_of_two_right (h : is_primitive_root ζ 2) : ζ = -1 :=
+begin
+  apply (eq_or_eq_neg_of_pow_two_eq_pow_two ζ 1 _).resolve_left,
+  { rw [← pow_one ζ], apply h.pow_ne_one_of_pos_of_lt; dec_trivial },
+  { simp only [h.pow_eq_one, one_pow] }
+end
+
+end integral_domain
+
+section integral_domain
+
+variables {ζ : units R} (h : is_primitive_root ζ k)
+
+protected
+lemma mem_roots_of_unity {n : ℕ+} (h : is_primitive_root ζ n) : ζ ∈ roots_of_unity n R :=
+h.pow_eq_one
+
+/-- The (additive) monoid equivalence between `zmod k`
+and the powers of a primitive root of unity `ζ`. -/
+def zmod_equiv_gpowers (h : is_primitive_root ζ k) : zmod k ≃+ additive (subgroup.gpowers ζ) :=
```

Comment:
Would it make more sense to state this in the other direction? I'm not sure which is more useful in the end, but the other way is "simpler."

Reply:
I think it doesn't matter too much. It's sort of awkard anyway with the `additive`.
Stating it the other way would introduce a `.symm` in the definition, and hence make the simp-lemmas below a bit harder as well.
So I would prefer to keep it like this.